moveit_core/online_signal_smoothing/src/acceleration_filter.cpp

```diff
@@ -0,0 +1,347 @@
+  size_t idx = 0;
+  for (const auto& joint_bound : joint_bounds)
+  {
+    for (const auto& variable_bound : *joint_bound)
+    {
+      const auto& target_accel = accelerations(idx);
+      if (variable_bound.acceleration_bounded_ && target_accel != 0.0)
+      {
+        // Find the ratio of clamped acceleration to original acceleration
+        const auto bounded_vel =
+            std::clamp(target_accel, variable_bound.min_acceleration_, variable_bound.max_acceleration_);
+        double joint_scaling_factor = bounded_vel / target_accel;
+        min_scaling_factor = std::min(min_scaling_factor, joint_scaling_factor);
+      }
+      ++idx;
+    }
+  }
+
+  return min_scaling_factor;
+}
+
+inline bool update_data(const OSQPDataWrapperPtr& data, OSQPWorkspace* work, Eigen::SparseMatrix<double>& A_sparse,
+                        const Eigen::VectorXd& lower_bound, const Eigen::VectorXd& upper_bound)
+{
+  data->update_A(work, A_sparse);
+  size_t num_constraints = A_sparse.rows();
+  data->u.block(0, 0, num_constraints - 1, 1) = upper_bound;
+  data->l.block(0, 0, num_constraints - 1, 1) = lower_bound;
+  data->u[num_constraints - 1] = ALPHA_UPPER_BOUND;
+  data->l[num_constraints - 1] = ALPHA_LOWER_BOUND;
+  return 0 == osqp_update_bounds(work, data->l.data(), data->u.data());
+}
+
+bool AccelerationLimitedPlugin::doSmoothing(Eigen::VectorXd& positions, Eigen::VectorXd& velocities,
+                                            Eigen::VectorXd& /* unused */)
+{
+  const size_t num_positions = velocities.size();
+  if (num_positions != num_joints_)
+  {
+    RCLCPP_ERROR_THROTTLE(
+        getLogger(), *node_->get_clock(), 1000,
+        "The length of the joint positions parameter is not equal to the number of joints, expected %zu got %zu.",
+        num_joints_, num_positions);
+    return false;
+  }
+  else if (last_positions_.size() != positions.size())
+  {
+    RCLCPP_ERROR_THROTTLE(getLogger(), *node_->get_clock(), 1000,
+                          "The length of the last joint positions not equal to the current, expected %zu got %zu. Make "
+                          "sure the reset was called.",
+                          last_positions_.size(), positions.size());
+    return false;
+  }
+
+  // formulate quadratic program to find the best new reference point subject to the robot's acceleration limits
+  // p_c: robot's current position
+  // v_c: robot's current velocity
+  // p_t: robot's target position
+  // acc: acceleration to be applied
+  // p_n: next position
+  // dt: time step
+  // p_n_hat: parameterize solution to be along the line from p_c to p_t
+  // p_n_hat = p_t*alpha + p_c*(1-alpha)
+  // define constraints
+  // p_c + v_c*dt + acc_min*dt^2 < p_n_hat < p_c + v_c*dt + acc_max*dt^2
+  // p_c + v_c*dt -p_t + acc_min*dt^2 < (p_c-p_t)alpha < p_c + v_c*dt -p_t + acc_max*dt^2
+  // 0 < alpha < 1
+  // define optimization
+  // opt ||alpha||
+  // s.t. constraints
+  // p_n = p_t*alpha + p_c*(1-alpha)
```

Comment:
This is dense, but also pretty important! Maybe this is worth elevating higher up in the overall description and/or README?

Reply:
I got a comment from @sjahr about documenting the math steps. I did add more information in the README.